File: SEM5/MobileApp/lab2_native/app/src/main/java/com/example/ezpensetracker/ExpenseService.kt
package com.example.ezpensetracker

import Expense

class ExpenseService {
    private val expenses = mutableListOf(
        Expense(1, 2000.00, "Spending", "Food", "31/10/2024", "The food for the entire month. I passed the date as the end of the specific month"),
        Expense(2, 200.00, "Spending", "Gym", "14/10/2024", "The gym subscription"),
        Expense(3, 3795.00, "Receiving", "Salary", "10/10/2024", ""),
        Expense(4, 660.00, "Receiving", "Meal tickets", "05/10/2024", "Meal tickets for last month")
    )

    fun getExpenses(): List<Expense> {
        return expenses.toList()
    }

//    fun addExpense(expense: Expense) {
//        var maxId = 0
//
//        for (e in expenses) {
//            if (e.id > maxId) {
//                maxId = e.id
//            }
//        }
//
//        val nextId = maxId + 1
//
//        val expenseWithId = expense.copy(id = nextId)
//
//        expenses.add(expenseWithId)
//    }
//
//
//    fun updateExpense(updatedExpense: Expense): Boolean {
//        val index = expenses.indexOfFirst { it.id == updatedExpense.id }
//        return if (index != -1) {
//            expenses[index] = updatedExpense
//            true  // Successfully updated
//        } else {
//            false  // Expense not found
//        }
//    }
//
//    fun deleteExpense(expenseId: Int): Boolean {
//        val index = expenses.indexOfFirst { it.id == expenseId }
//        return if (index != -1) {
//            expenses.removeAt(index)
//            true  // Successfully deleted
//        } else {
//            false  // Expense not found
//        }
//    }
//
//    fun getExpenseById(id: Int): Expense? {
//        return expenses.find { it.id == id }
//    }
}

File: SEM5/MobileApp/lab3/expense_tracker_ui/lib/services/ExpenseService.dart
import 'package:expense_tracker_ui/domain/Expense.dart';

class ExpenseService {
  final List<Expense> _expenses = [
    Expense(id: 1, amount: 3725.00, category: "Salary", date: "10.10.2024", type: "RECEIVING", notes: ""),
    Expense(id: 2, amount: 660.00, category: "Meal Tickets", date: "05.10.2024", type: "RECEIVING", notes: "Meal tickets received as benefit for job"),
    Expense(id: 3, amount: 180.00, category: "Gym", date: "18.10.2024", type: "SPENDING", notes: "gym subscription paid automatically for September"),
  ];

  void addExpense(Expense expense) {
    _expenses.add(expense);
  }

  void removeExpense(int id) {
    _expenses.removeWhere((expense) => expense.id == id);
  }

  Expense? getExpenseById(int id) {
    return _expenses.firstWhere((expense) => expense.id == id, orElse: () => Expense(id: -1, amount: 0, category: "", date: "", type: "", notes: ""));
  }

  void updateExpense(Expense updatedExpense) {
    final index = _expenses.indexWhere((expense) => expense.id == updatedExpense.id);
    if (index != -1) {
      _expenses[index] = updatedExpense;
    }
  }

  List<Expense> getAllExpenses() {
    return List.unmodifiable(_expenses);
  }
}
